### PR TITLE
[quantization] State dict serialization of nnq.Linear

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24117 [quantization] JIT serialization for Conv2d
* #24116 [quantization] state_dict serialization for Conv2d + some bugfixes
* #24115 [quantization] Re-work Conv2d
* #24048 [quantization] JIT Serialization of nnq.Linear
* **#24047 [quantization] State dict serialization of nnq.Linear**
* #24046 [quantization] Simplified nnq.Linear class

Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Differential Revision: [D16728346](https://our.internmc.facebook.com/intern/diff/D16728346)

### CODE REVIEW

File: torch/nn/quantized/modules/linear.py [hunk 1]
```diff
@@ -118,6 +119,45 @@ def forward(self, x):
         return torch.ops.quantized.fbgemm_linear(
             x, self._packed_weight, self.bias, self.scale, self.zero_point)
 
+    # ===== Serialization methods =====
+    # The special consideration here is that we have to unpack the weights into their
+    # regular QTensor form for serialization. Packed weights should not live
+    # outside the process in which they were created, rather they should be derived
+    # from the QTensor weight.
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: does `super()` work in python2?

File: torch/nn/quantized/modules/linear.py [hunk 2]
```diff
@@ -118,6 +119,45 @@ def forward(self, x):
...
+        state_dict.pop(prefix + 'zero_point')
+        super()._load_from_state_dict(state_dict, prefix, local_metadata, False,
+                                      missing_keys, unexpected_keys, error_msgs)
+
+    # Function rather than property to make sure that JIT serialization doesn't
+    # register this as an attribute
+    def weight(self):
+        return torch.ops.quantized.fbgemm_linear_unpack(self._packed_weight)
+
+    def set_weight(self, w):
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: add a test for set_weight function as well? I think this is should be used in from_float

Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: line 188 we can probably use set_weight

File: torch/nn/quantized/modules/linear.py (hunk 1 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: nope, good catch

**Review by raghuramank100 — APPROVED** [review]

File: torch/nn/quantized/modules/linear.py [hunk 3]
```diff
@@ -110,14 +111,56 @@ def __init__(self, in_features, out_features, bias_=True):
...
+            self.bias.copy_(state_dict[prefix + 'bias'])
+            state_dict.pop(prefix + 'bias')
+
+        self.scale = float(state_dict[prefix + 'scale'])
+        state_dict.pop(prefix + 'scale')
+
+        self.zero_point = int(state_dict[prefix + 'zero_point'])
+        state_dict.pop(prefix + 'zero_point')
+
+        super()._load_from_state_dict(state_dict, prefix, local_metadata, False,
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: ..another one here.

**Review by jerryzh168 — APPROVED** [review]
LGTM, please fix the extra super()

File: torch/nn/quantized/modules/linear.py (hunk 3 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: I'm surprised flake8 doesn't catch these in the pre-commit hook...

**Pull request closed without merging**